webtransport-quinn/src/session.rs

```diff
@@ -100,16 +102,47 @@ impl Session {
         Ok((SendStream::new(send), RecvStream::new(recv)))
     }
 
-    pub async fn read_datagram(&self) {
-        unimplemented!("datagrams")
+    /// Asynchronously receives an application datagram from the remote peer.
+    ///
+    /// This method is used to receive an application datagram sent by the remote
+    /// peer over the connection.
+    /// It waits for a datagram to become available and returns the received [`Datagram`].
+    /// 
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use webtransport_quinn::Session;
+    /// # use quinn::Connecting
+    /// # use anyhow::Result;
+    /// # async fn run(connection: Connecting) -> Result<()> {
+    /// # let conn = &connection.await?;
+    /// # let req = webtransport_quinn::accept(conn.clone()).await?;
+    /// # let session = request.ok().await?;
+    /// let datagram = session.read_datagram().await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn read_datagram(&self) -> Result<Datagram, ErrorCode> {
```

Comment:
This should be:

```rust
pub async fn read_datagram(&self) -> Result<Bytes, SessionError> {
```

1. We don't want to return the QUIC stream ID to the user; it's hidden on purpose. You can validate that it matches Session ID if you want.
2. That means we can return `Bytes` directly. [Just like Quinn](https://docs.rs/quinn/latest/quinn/struct.ReadDatagram.html), except without using a custom Future.
3. No need for a custom error `ErrorCode` for datagrams since it only removes error information. Use `SessionError` like everything else.